Fix: different frame size as inputs of the filter graph

valnoel opened the pull request:
In order to mux audio channels, AvTranscoder uses the [amerge](https://ffmpeg.org/ffmpeg-filters.html#amerge-1) filter.
If, as inputs of the filter graph containing this _amerge_ filter, we put frames with different sizes, the resulting frame size seems to be the minimum of the input frame size. This means that one or more output channels can have cut frames, changing the reading speed and generating clicks into the final audio stream.

Using the internal filter buffer, the output frame is ensured not to lose audio samples, solving the problem of speed and clicks.
However, the filter internal buffer is quiet limited and quickly overflowed, making the process crashing before the end of the program. So, using custom frame buffers before the filter graph inputs fixes this other problem.

Commit:
FilterGraph: push input frames into the filter buffer

Let the filter managing the output frame size, using its internal
frame buffer queue.

Commit:
FilterGraph: add new internal FrameBuffer class

Used as frame buffers on the filter graph inputs.
For the moment, this first version only handles audio frames.

Commit:
FilterGraph: use FrameBuffers for each filter graph input

Returning frames with the same size as inputs of the graph,
avoiding to overflow the internal filter buffers.

Commit:
FilterGraph: improve input frame size computing

Commit:
FilterGraph: frame buffers can be bypassed if not necessary

If the input frames have the same size and the buffers are empty, bypass
to avoid useless data copies.

cchampet commented on src/AvTranscoder/filter/FilterGraph.hpp:
```diff
@@ -5,15 +5,60 @@
+private:
+    void popFrame();
+
+    const AudioFrameDesc _audioFrameDesc;
```
It seems that you have a compilation error because this attribute is _const_...

valnoel replied:
Yes, thanks! Now, it's fixed in https://github.com/avTranscoder/avTranscoder/pull/303/commits/380f2382f2e0c2073aa2a00b2971f32abbe89d1e

Commit:
FilterGraph: minor clean into FrameBuffer class

Commit:
FilterGraph: refactoring of getMinInputFrameSize() private method

Add a new getAvailableFrameSize() private method

Commit:
StreamTranscoder: flush filter graph buffers after decoding process

Add utility methods, documentation and logs into FilterGraph
and FrameBuffer classes

Commit:
StreamTranscoder: reset frame that has not been decoded

Avoid filling the filter graph buffers with wrong data frames.
Improve documentation into FilterGraph::FrameBuffer class.

Commit:
FilterGraph: renames FrameBuffer to AudioFrameBuffer

Since video frames cannot be split, this filter graph buffers
will be specialized to audio frames.

Commit:
FilterGraph: remove useless areInputAudioFrames() method

Replace it by a simple AudioFrameBuffer vector empty() test, since
buffers are initialized only for audio frames (into the addInBuffer()
method)
And rename _inputAudioFramesBuffer attribute to _inputAudioFrameBuffers

Commit:
FilterGraph: fix AudioFrameBuffer class name case typo

Commit:
FilterGraph: fix AudioFrameBuffers order

Commit:
FilterGraph: fix audio frames management with different sample formats

Comparing the frame sizes in samples (instead of bytes)

Commit:
Test: add a new python test for audio channels muxing

Audio channels are extracted from different file format sources, to
stereo and 5.1 output audio streams.

valnoel commented:
@cchampet Ready to merge?

Commit:
FilterGraph: fixes Python binding compilation on Windows

Export AudioFrameBuffer class symbol for DLL

cchampet commented on test/pyTest/testMuxAudioChannels.py:
```diff
@@ -0,0 +1,101 @@
+    dst_inputFile = av.InputFile(outputFileName)
+    dst_audioProperties = dst_inputFile.getProperties().getAudioProperties()
+    assert_equals(1, len(dst_audioProperties))
+    assert_equals(2, dst_audioProperties[0].getNbChannels())
```
Could you check the duration of the output stream?

valnoel replied:
Actually, it was already done using the process statistics, but it is indeed better using the output file properties: 9abcb90

cchampet commented on test/pyTest/testMuxAudioChannels.py:
```diff
@@ -0,0 +1,101 @@
+    dst_inputFile = av.InputFile(outputFileName)
+    dst_audioProperties = dst_inputFile.getProperties().getAudioProperties()
+    assert_equals(1, len(dst_audioProperties))
+    assert_equals(6, dst_audioProperties[0].getNbChannels())
```
Could you check the duration of the output stream?

valnoel replied:
Duration check using the output file properties: 9abcb90

cchampet commented on src/AvTranscoder/transcoder/StreamTranscoder.cpp:
```diff
@@ -569,7 +571,26 @@ bool StreamTranscoder::processTranscode()
+            if(!_filterGraph->hasFilters() || !_filterGraph->hasBufferedFrames(index))
+            {
+                continueProcess = false;
+                continue;
```
Did you consider to user __break__ instead of __continue__? Because in this case, do we need to continue to get the remaining buffers in the filter graph?

valnoel replied:
Sure, it is better! See 0e49751

Commit:
Test: check output files duration directly through file properties

Into audio channels muxing tests

Commit:
StreamTranscoder: avoid checking every filter graph buffer if the process won't be continued

valnoel commented:
@cchampet  RTM?

cchampet commented on src/AvTranscoder/filter/FilterGraph.cpp:
```diff
@@ -16,6 +15,123 @@ extern "C" {
+
+    // Concatenate frames data
+    size_t extractedDataSize = 0;
+    unsigned char* outputData = new unsigned char[size];
```
Is this type mean that the supported data type is only 1byte per sample?

valnoel replied:
No, it is actually the raw data in bytes, without considering the sample format or whatever.

cchampet commented:
@valnoel I had only one question.
Except that, you can merge whenever you want ;)

valnoel commented:
@cchampet OK then! :)
